dev/ese/src/sync/sync.cxx

```diff
@@ -4303,6 +3817,323 @@ void CKernelSemaphore::Release( const INT cToRelease )
+
+    State().IncWait();
+    OSSYNC_FOREVER
+    {
+        //  read the current state of the semaphore
+
+        const CSemaphoreState stateCur = (CSemaphoreState&) State();
+
+        //  see if we have an available count
+
+        if ( stateCur.CAvail() > 0 )
+        {
+            //  we ourselves are waiting on the semaphore, so there had better be at least
+            //  one waiter in the state
+
+            OSSYNCAssert( stateCur.CWait() > 0 );
+
+            //  try to atomically acquire the semaphore and decrement the number of waiters
+
+            const CSemaphoreState stateNew( stateCur.CAvail() - 1, stateCur.CWait() - 1 );
+
+            if ( State().FChange( stateCur, stateNew ) )
+            {
+                //  the transaction succeeded, return success
+
+                return fTrue;
+            }
+        }
+        else if ( dwRemaining == 0 )
+        {
+            //  we were unable to acquire the semaphore in the time permitted
+
+            break;
+        }
+        else
+        {
+            //  check and update the remaining time
+
+            const DWORD dwElapsed = DwOSSyncITickTime() - dwStart;
+
+            if ( dwElapsed > dwTimeout )
+            {
+                //  we were unable to acquire the semaphore in the time permitted
+
+                break;
+            }
+
+            dwRemaining = dwTimeout - dwElapsed;
+
+            //  wait for the semaphore counter value to change
+            //
+            //  NOTE:  we may have exited the wait due to a spurious OS level wake up,
+            //  so the state machine ensures that the available count is re-checked on
+            //  each iteration
+
+            if ( !_FWait( stateCur.CAvail(), dwRemaining ) )
+            {
+                //  we were unable to acquire the semaphore in the time permitted
+
+                break;
+            }
+        }
+    }
+    State().DecWait();
+
+    return fFalse;
+}
+
+//  releases all waiters on the semaphore
+
+void CSemaphore::ReleaseAllWaiters()
+{
+    //  try forever until we successfully change the state of the semaphore
+
+    OSSYNC_FOREVER
+    {
+        //  read the current state of the semaphore
+
+        const CSemaphoreState stateCur = (CSemaphoreState&) State();
+
+        if ( stateCur.CAvail() > 0 && stateCur.CWait() > 0 )
+        {
+            //  the existing waiters are in transition, retry
+
+            continue;
+        }
+        else if ( stateCur.CWait() <= 0 )
+        {
+            //  there are no waiters, we're done
+
+            return;
+        }
+        else
+        {
+            //  attempt to change the semaphore to have an available count equal
+            //  to the number of waiters
+
+            const CSemaphoreState stateNew( stateCur.CWait(), stateCur.CWait() );
+
+            if ( State().FChange(stateCur, stateNew ) )
+            {
+                volatile void *pv = State().PAvail();
```

Comment:
> On the other hand, it seems improper for CSemaphoreState to be waiting and waking...

I agree with that: it's the semaphore that should be waiting for its state to change, not the state itself.

As for the problem itself, I committed a reworked version of that code in [161b0bcb](https://github.com/evgenykotkov/Extensible-Storage-Engine/commit/161b0bcbed4fd9751e0c4576dcc5c88edfb673ec). The new version (hopefully) breaks the dependency between `PAvail()` and `CAvail()` by using a typed pointer for the address that we wait on.  We also make the `CompareAddress` point to the value of the same type and use a `static_assert<>` for the sizes of two values, just in case.

How does that sound?